gdal/ogr/ogrsf_frmts/geojson/directedacyclicgraph.hpp

```diff
@@ -0,0 +1,253 @@
+#include <vector>
+
+#include <cassert>
+
+namespace gdal
+{
+
+// See https://en.wikipedia.org/wiki/Directed_acyclic_graph
+template<class T, class V = std::string> class DirectedAcyclicGraph
+{
+    std::set<T> nodes;
+    std::map<T, std::set<T>> incomingNodes; // incomingNodes[j][i] means an edge from i to j
+    std::map<T, std::set<T>> outgoingNodes; // outgoingNodes[i][j] means an edge from i to j
+    std::map<T, V> names;
+
+public:
+    DirectedAcyclicGraph() = default;
+
+    void addNode(const T& i, const V& s) { nodes.insert(i); names[i] = s; }
+    void removeNode(const T& i);
+    const char* addEdge(const T& i, const T& j);
+    const char* removeEdge(const T& i, const T& j);
+    bool isTherePathFromTo(const T& i, const T& j) const;
+    std::vector<T> findStartingNodes() const;
+    std::vector<T> getTopologicalOrdering();
+};
+
+template<class T, class V>
+void DirectedAcyclicGraph<T, V>::removeNode(const T& i)
+{
+    nodes.erase(i);
+    names.erase(i);
+
+    {
+        auto incomingIter = incomingNodes.find(i);
+        if( incomingIter != incomingNodes.end() )
+        {
+            for( const T& j: incomingIter->second )
+            {
+                auto outgoingIter = outgoingNodes.find(j);
+                assert(outgoingIter != outgoingNodes.end());
+                auto iterJI = outgoingIter->second.find(i);
+                assert(iterJI != outgoingIter->second.end());
+                outgoingIter->second.erase(iterJI);
+                if( outgoingIter->second.empty() )
+                    outgoingNodes.erase(outgoingIter);
+            }
+            incomingNodes.erase(incomingIter);
+        }
+    }
+
+    {
+        auto outgoingIter = outgoingNodes.find(i);
+        if( outgoingIter != outgoingNodes.end() )
+        {
+            for( const T& j: outgoingIter->second )
+            {
+                auto incomingIter = incomingNodes.find(j);
+                assert(incomingIter != incomingNodes.end());
+                auto iterJI = incomingIter->second.find(i);
+                assert(iterJI != incomingIter->second.end());
+                incomingIter->second.erase(iterJI);
+                if( incomingIter->second.empty() )
+                    incomingNodes.erase(incomingIter);
+            }
+            outgoingNodes.erase(outgoingIter);
+        }
+    }
+}
+
+template<class T, class V>
+const char* DirectedAcyclicGraph<T, V>::addEdge(const T& i, const T& j)
+{
+    if( i == j )
+    {
+        return "self cycle";
+    }
+    const auto iterI = outgoingNodes.find(i);
+    if( iterI != outgoingNodes.end() &&
+        iterI->second.find(j) != iterI->second.end() )
+    {
+        return "already inserted edge";
+    }
+
+    if( nodes.find(i) == nodes.end() )
+    {
+        return "node i unknown";
+    }
+    if( nodes.find(j) == nodes.end() )
+    {
+        return "node j unknown";
+    }
+
+    if( isTherePathFromTo(j, i) )
+    {
+        return "can't add edge: this would cause a cycle";
```

Comment:
should you drop both instead and sort these two by some other means? this line will keep the attribute order feature-order-dependant still.

Reply:
good point, but what would you drop if we have edges A->B and B->C, and we now want to insert C->A ?

Reply:
all three as they can't be treated as definitions? but you can't do that on this same pass as then next feature will push them back. you generally need to build cyclic graph and then cut off cycles explicitly.

other strategies that come to mind:
 - push the edges into a set first, order it alphabetically as tuples, push into current dag. will keep alphabetical ordering in case of conflicts.
 - count number of occurences for each edge, sort by that then alphabetically, push into current dag. will keep most common order in case of conflicts.
  - say "it will be this way" and do nothing, maybe put a note. current one still works. :)

Reply:
I guess first option can also be simplified to "edge with from<to beats all conflicting edges with from>to". So in your example C->A is dropped in any order of insertion.

Reply:
I believe I will keep suggested improvements (thanks for them) for a next time. Conflicting order is a less common case, and cannot be generated with GDAL.